### PR TITLE
[SPARK-39338][SQL] Remove dynamic pruning subquery if pruningKey's references is empty

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

Remove dynamic pruning subquery if pruningKey's references is empty. For example:
```sql
SELECT f.store_id,
       f.date_id,
       s.state_province
FROM (SELECT   store_id,
               date_id,
               product_id
      FROM   fact_stats
      WHERE  date_id <= 1000
      UNION ALL
      SELECT 4 AS store_id,
               date_id,
               product_id
      FROM   fact_sk
      WHERE  date_id >= 1300) f
JOIN dim_store s
ON f.store_id = s.store_id
WHERE s.country IN ('US', 'NL')
```

Before this PR | After this PR
-- | --
image | image

### Why are the changes needed?

Remove useless dynamic pruning subquery because it can't reduce partition.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by wangyum** [comment]
cc @cloud-fan

**Review by cloud-fan — APPROVED** [review]
LGTM, cc @sigmod

File: sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/CleanupDynamicPruningFilters.scala [hunk 1]
```diff
@@ -54,7 +54,8 @@ object CleanupDynamicPruningFilters extends Rule[LogicalPlan] with PredicateHelp
         val newCondition = condition.transformWithPruning(
           _.containsPattern(DYNAMIC_PRUNING_SUBQUERY)) {
           case dynamicPruning: DynamicPruningSubquery
-              if unnecessaryPruningKeys.contains(dynamicPruning.pruningKey) =>
+              if unnecessaryPruningKeys.contains(dynamicPruning.pruningKey) ||
+                dynamicPruning.pruningKey.references.isEmpty =>
```
Comment (weixiuli) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/CleanupDynamicPruningFilters.scala:58]: Can we set  the _unnecessaryPruningKeys_ to a lazy and  check  the `dynamicPruning.pruningKey.references.isEmpty` before `  the unnecessaryPruningKeys.contains(dynamicPruning.pruningKey)` ?

**Review by weixiuli — APPROVED** [review]
+1 LGTM.

**Review by sigmod — APPROVED** [review]

**Comment by wangyum** [comment]
Merged to master.

**Pull request closed without merging**